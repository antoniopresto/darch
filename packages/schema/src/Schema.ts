import { RuntimeError } from '@darch/utils/lib/RuntimeError';
import { StrictMap } from '@darch/utils/lib/StrictMap';
import { expectedType } from '@darch/utils/lib/expectedType';
import { getTypeName } from '@darch/utils/lib/getTypeName';
import { invariantType } from '@darch/utils/lib/invariant';
import { simpleObjectClone } from '@darch/utils/lib/simpleObjectClone';
import { ForceString } from '@darch/utils/lib/typeUtils';

import { SchemaDefinitionInput } from './TSchemaConfig';
import { AnyParsedSchemaDefinition, ParsedSchemaDefinition, TypeFromSchema } from './TSchemaParser';
import { parseValidationError, ValidationCustomMessage } from './applyValidator';
import { parseSchemaFields } from './getSchemaErrors';
import { parseSchemaDefinition } from './parseSchemaDefinition';
import type { ObjectTypeComposer } from 'graphql-compose';
import { isBrowser } from '@darch/utils/lib/isBrowser';
import { GraphQLInputObjectType, GraphQLObjectType } from 'graphql';
import { areEqual } from '@darch/utils/lib/areEqual';

export { RuntimeError } from '@darch/utils/lib/RuntimeError';
export * from './parseSchemaDefinition';
export * from './schemaInferenceUtils';

export class Schema<DefinitionInput extends SchemaDefinitionInput> {
  _infer!: TypeFromSchema<DefinitionInput>;

  private readonly __definition: any;

  get definition(): ParsedSchemaDefinition<DefinitionInput> {
    return this.__definition;
  }

  _description = '';

  constructor(schemaDef: DefinitionInput) {
    this.__definition = parseSchemaDefinition(schemaDef);
  }

  parse(
    input: any,
    options?: { partial?: boolean; customMessage?: ValidationCustomMessage }
  ): TypeFromSchema<DefinitionInput> {
    const { customMessage } = options || {};
    const { errors, parsed } = this.safeParse(input, options?.partial);

    if (errors.length) {
      const err: any = parseValidationError(input, customMessage, errors.join(' \n'));
      err.isSchemaValidationError = true;
      err.fieldErrors = errors;
      throw err;
    }

    return parsed as any;
  }

  validate(input: any): input is TypeFromSchema<DefinitionInput> {
    try {
      this.parse(input);
      return true;
    } catch (e) {
      return false;
    }
  }

  safeParse(input: any, partial?: boolean): { errors: string[]; parsed: unknown } {
    const SchemaConstructor: any = this.constructor;

    const errors: string[] = [];
    const parsed: any = {};

    if (!input || typeof input !== 'object' || Array.isArray(input)) {
      throw new RuntimeError(`Invalid input. Expected object, found ${getTypeName(input)}`, {
        input,
      });
    }

    Object.keys(this.definition).forEach((currField) => {
      const value = input[currField];
      const definition = this.definition[currField];

      if (value === undefined && partial) {
        return;
      }

      const result = parseSchemaFields({
        createSchema: (def) => new SchemaConstructor(def),
        fieldName: currField,
        definition,
        value,
      });

      parsed[currField] = result.parsed;
      errors.push(...result.errors);
    });

    return { parsed, errors };
  }

  describe(...descriptions: [comment: string] | [{ [K in keyof DefinitionInput]?: string }]): this {
    if (descriptions.length === 1 && typeof descriptions[0] === 'string') {
      this._description = descriptions[0];
      return this;
    }

    const commentsConfig = descriptions[0];

    invariantType({ commentsConfig }, 'object', { commentsConfig });

    const definition: AnyParsedSchemaDefinition = this.definition;

    Object.entries(commentsConfig).forEach(([name, comment]) => {
      invariantType({ [name]: definition[name] }, 'object', `"${name}" is not in schema definition.`);
      definition[name].description = comment || '';
    });

    return this;
  }

  removeField<K extends ForceString<keyof DefinitionInput>>(field: K | K[]): Schema<OmitDefinitionFields<this, K>> {
    const fields: string[] = Array.isArray(field) ? field : [field];
    const clone = this.clone();

    for (const k in clone.__definition) {
      if (fields.includes(k)) {
        delete clone.__definition[k];
      }
    }

    return clone as any;
  }

  addFields<T extends SchemaDefinitionInput>(definition: T): Schema<ExtendDefinition<this, T>> {
    return this.clone(definition) as any;
  }

  get __isTaskforceSchema() {
    return true;
  }

  clone(): this;
  clone<T extends SchemaDefinitionInput>(extend: T): ExtendDefinition<this, T>;
  clone(...args: any[]) {
    const def = simpleObjectClone({ ...this.definition, ...args[0] });
    return createSchema(def) as any;
  }

  private __id: string | null = null;
  get id() {
    return this.__id;
  }

  identify<ID extends string>(id: ID): this & { id: ID } {
    if (id && id === this.id) return this as any;

    if (this.id) {
      throw new Error(
        `Trying to replace existing id "${this.id}" with "${id}". You can clone it to create a new Schema.`
      );
    }

    expectedType({ id }, 'string', 'truthy');

    if (Schema.register.has(id) && Schema.register.get(id) !== this) {
      if (areEqual(Schema.register.get(id), this)) {
        return this as any;
      } else {
        throw new Error(`Schema with id "${id}" already registered.`);
      }
    }

    this.__id = id;
    Schema.register.set(id, this);

    return this as any;
  }

  static register = new StrictMap();

  private _otc: ObjectTypeComposer | undefined;
  otc = (): ObjectTypeComposer => {
    if (this._otc) return this._otc;
    if (isBrowser() || typeof module?.require !== 'function') {
      throw new Error('GraphQL transformation is not available on browser.');
    }

    if (!this.id) {
      throw new Error(
        'Should schema.identify() before converting to Graphql.' +
          '\nYou can call schema.clone() to choose a different identification.'
      );
    }

    const { schemaToGQL } = module
      // preventing build tools transforming require calls
      .require('./schemaToGQL');
    //
    return (this._otc = schemaToGQL(this.id, this.definition));
  };

  private _graphqlType: GraphQLObjectType | undefined;
  graphqlType = (): any => {
    if (this._graphqlType) return this._graphqlType;
    return (this._graphqlType = this.otc().getType());
  };

  private _graphqlInputType: GraphQLInputObjectType | undefined;
  graphqlInputType = (): any => {
    if (this._graphqlInputType) return this._graphqlInputType;
    return (this._graphqlInputType = this.otc().getInputType());
  };
}

export const DarchSchema = Schema;

export function createSchema<DefinitionInput extends Readonly<SchemaDefinitionInput>>(
  fields: DefinitionInput
): Schema<DefinitionInput>;

export function createSchema<DefinitionInput extends Readonly<SchemaDefinitionInput>>(
  name: string,
  fields: DefinitionInput
): Schema<DefinitionInput>;

export function createSchema<DefinitionInput extends Readonly<SchemaDefinitionInput>>(
  ...args: [string, DefinitionInput] | [DefinitionInput]
): Schema<DefinitionInput> {
  const fields = args.length === 2 ? args[1] : args[0];
  const id = args.length === 2 ? args[0] : undefined;

  const schema = new Schema<DefinitionInput>(fields);
  if (id) return schema.identify(id);

  return schema;
}

export { createSchema as createType };

type OmitDefinitionFields<T, Keys extends string> = T extends { definition: Record<string, any> }
  ? Omit<T['definition'], Keys>
  : never;

type ExtendDefinition<T, Ext extends SchemaDefinitionInput> = T extends { definition: Record<string, any> }
  ? {
      [K in keyof (T['definition'] & Ext)]: (T['definition'] & Ext)[K];
    }
  : never;
