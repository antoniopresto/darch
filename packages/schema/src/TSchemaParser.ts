import { OnlyKnown } from '@darch/utils/dist/typeUtils';

import { FieldType } from './FieldType';
import { CommonFieldConfig, FieldDefinitionConfig } from './TSchemaConfig';
import { CursorType } from './fields/CursorField';
import { FieldTypeName, FieldTypes } from './fields/fieldTypes';
import { AnyStringFieldDefinition, ParseStringDefinition } from './parseStringDefinition';
import { InferRecordFieldType } from './fields/RecordField';

export type AnyParsedFieldDefinition = {
  [K in keyof FieldTypes]: {
    type: K;
    def?: any; // TODO infer
    list?: boolean;
    optional?: boolean;
    description?: string;
  };
}[keyof FieldTypes];

export function isSchemaAsFieldDef(input: any): input is AnyParsedFieldDefinition & {
  type: 'schema';
  def: AnyParsedSchemaDefinition;
} {
  return input?.type === 'schema' && !!input.def && typeof input.def === 'object';
}

export type AnyParsedSchemaDefinition = {
  [K: string]: AnyParsedFieldDefinition;
};

export type ParsedSchemaDefinition<T> = {
  [K in keyof T]: ParsedFieldDefinition<T[K]>;
};

export type Infer<T> = TypeFromSchema<T>;

export type TypeFromSchema<T> =
  // Z - ParsedFieldDefinition -> Optional<List<GetType<T>>>
  // A - { definition } -> iterate definition -> TypeFromFieldDef
  // B - { type: { definition }} -> Optional<List<A>>
  // C - AnyStringFieldDefinition -> Optional<List<T>>
  // D - readonly [string, ...string[]] -> GetType<T>
  // E - FieldType<any, FieldTypeName, any> ->
  // F - Record<string, FieldDefinition>
  // G - { type: 'schema', def: Def }
  // H - {address: { schema: schemaDefinitionInput }, name: { string: {min: 2 }}}

  // ================ case Z START ================
  T extends {
    type: FieldTypeName;
    list?: infer List;
    optional?: infer Optional;
  }
    ? Maybe_List<GetTSType<ParsedFieldDefinition<T>>, List, Optional>
    : // ================ case Z END ================

    // ================ case A START ================
    T extends { definition: ParsedSchemaDefinition<any> }
    ? TypeFromSchemaDefinition<T['definition']>
    : // ================ case A END ==================

    // ================ case B START ================
    T extends {
        type: { definition: ParsedSchemaDefinition<any> };
        list?: infer List;
        optional?: infer Optional;
      }
    ? Maybe_List<TypeFromSchemaDefinition<T['type']['definition']>, List, Optional>
    : // ================ case B END ==================

    // ================ case C START ================
    T extends AnyStringFieldDefinition
    ? TypeFromSchema<ParseStringDefinition<T>>
    : // ================ case C END ==================

    // ================ case D START ================
    T extends readonly [string, ...string[]]
    ? TypeFromSchema<ParsedFieldDefinition<T>>
    : // ================ case D END ==================

    // ================ case E START ================
    T extends {
        __isFieldType: true;
        typeName: infer TypeName;
        isList?: infer List;
        isOptional?: infer Optional;
        def?: infer Def;
      }
    ? TypeFromSchema<
        ParsedFieldDefinition<{
          type: TypeName;
          optional: Optional;
          list: List;
          def: Def;
        }>
      >
    : // ================ case E END ==================

    // ================ case F START ================
    T extends { [K: string]: FieldDefinitionConfig }
    ? {
        -readonly [K in keyof T]: ParsedFieldDefinition<T[K]>;
      } extends infer Parsed
      ? Parsed extends Record<string, any>
        ? {
            // not a generic to better Intellisense debug
            [K in keyof ({
              [K in NullableKeys<Parsed, 'required'>]: TypeFromSchema<Parsed[K]>;
            } & {
              [K in NullableKeys<Parsed, 'nullable'>]?: TypeFromSchema<Parsed[K]>;
            })]: ({
              [K in NullableKeys<Parsed, 'nullable'>]?: TypeFromSchema<Parsed[K]>;
            } & {
              [K in NullableKeys<Parsed, 'required'>]: TypeFromSchema<Parsed[K]>;
            })[K];
          }
        : never
      : never
    : // ================ case F END ==================

    // ================ case G START ================
    T extends {
        type: 'schema';
        def: ParsedSchemaDefinition<any>;
        list?: infer List;
        optional?: infer Optional;
      }
    ? Maybe_List<
        {
          // not a generic to better Intellisense debug
          [K in keyof ({
            [K in NullableKeys<T['def'], 'required'>]: TypeFromSchema<T['def'][K]>;
          } & {
            [K in NullableKeys<T['def'], 'nullable'>]?: TypeFromSchema<T['def'][K]>;
          })]: ({
            [K in NullableKeys<T['def'], 'nullable'>]?: TypeFromSchema<T['def'][K]>;
          } & {
            [K in NullableKeys<T['def'], 'required'>]: TypeFromSchema<T['def'][K]>;
          })[K];
        },
        List,
        Optional
      >
    : // ================ case G END ==================

    // ================ case H START ================
    T extends {
        schema: infer Schema;
        optional?: infer Optional;
        list?: infer List;
      }
    ? T extends { type: string }
      ? never
      : Maybe_List<TypeFromSchema<Schema>, List, Optional>
    : // ================ case H START ================

    // ================ case H START ================
    [IsFieldAsSingleKey<T>] extends [true]
    ? TypeFromSchema<ParseFieldNameAsSingleKey<T>>
    : // ================ case H START ================
      // END
      never;

// const schema1 = createSchema({ x: ['a', 'b'] } as const)
// const schema2 = new Schema({ x: ['a', 'b'] } as const)
// // type Z = TypeFromSchema<ParsedFieldDefinition<['a', 'b']>>;
// type A = TypeFromSchema<typeof schema1>;
// type B = TypeFromSchema<{
//   type: typeof schema2;
//   list: true;
//   optional: true;
// }>;
// type C = TypeFromSchema<'[int]?'>;
// type D = TypeFromSchema<['a', 'b']>;
// type D2 = TypeFromSchema<readonly ['a', 'b']>;
// const s = { nameFromType: StringField.create().list().optional() } as const;
// type E = TypeFromSchema<typeof s>;

//  Z | AnyParsedFieldDefinition
//  A | AnySchema
//  B | { type: AnySchema; list?: boolean; optional?: boolean };
//  C | AnyStringFieldDefinition
//  D | readonly [string, ...string[]]
//  E | FieldType<any, FieldTypeName, any>

type HandleList<JSType, Is> = [Is] extends [true] ? JSType[] : JSType;
type HandleOptional<JSType, Is> = [Is] extends [true] ? JSType | undefined : JSType;
type Maybe_List<Type, IsList, IsOptional> = HandleOptional<HandleList<Type, IsList>, IsOptional>;

type GetTSType<T> =
  // ====== START enum type ======
  T extends { type: 'enum'; def: Array<infer En> } //
    ? En
    : T extends { type: 'enum'; def: Readonly<Array<infer En>> }
    ? En
    : // ====== END enum type ======

    // ====== START union type ======
    T extends { type: 'union'; def: Array<infer En> } //
    ? TypeFromSchema<En>
    : T extends { type: 'union'; def: Readonly<Array<infer En>> }
    ? TypeFromSchema<En>
    : // ====== END union type ======

    // ====== START record type ======
    T extends { type: 'record'; def: infer Def } //
    ? InferRecordFieldType<Def>
    : // ====== END record type ======

    T extends { type: 'int' | 'float' }
    ? number
    : T extends { type: 'string' | 'email' | 'ulid' }
    ? string
    : T extends { type: 'boolean' }
    ? boolean
    : T extends { type: 'date' }
    ? Date
    : T extends { type: 'cursor' }
    ? CursorType
    : T extends { type: 'schema'; def: infer T }
    ? TypeFromSchema<T>
    : T extends { type: 'unknown' }
    ? unknown
    : T extends { type: 'undefined' }
    ? undefined
    : T extends { type: 'any' }
    ? any
    : never;

export type ParsedFieldDefinition<T> =
  //
  // ===== start union type ====
  (
    T extends readonly [readonly FieldDefinitionConfig[]]
      ? {
          type: 'union';
          optional: HasNullableDefinition<T[0][number]>;
          list: false;
          def: T[0];
          description?: string;
        }
      : T extends {
          type: 'union';
          def: readonly FieldDefinitionConfig[] | FieldDefinitionConfig[];
          optional?: infer Optional;
          list?: infer List;
          description?: infer Description;
        }
      ? {
          type: 'union';
          optional: [Optional] extends [true] ? true : HasNullableDefinition<T['def'][number]>;
          list: [List] extends [true] ? true : false;
          def: T['def'];
          description?: Description;
        }
      : T extends FieldType<any, 'union', readonly any[] | any[]>
      ? {
          type: 'union';
          optional: [T['isOptional']] extends [true] ? T['isOptional'] : HasNullableDefinition<T['def'][number]>;
          list: T['isList'] extends true ? true : false;
          def: T['def'];
          description?: string;
        }
      : // ==== end union type ====

      T extends readonly [string, ...string[]]
      ? {
          type: 'enum';
          def: T;
          list: false;
          optional: false;
          description?: string;
        }
      : T extends { definition: infer SchemaDef }
      ? {
          type: 'schema';
          def: SchemaDef;
          list: false;
          optional: false;
          description?: string;
        }
      : // start 'schema'
      T extends {
          type: { definition: infer SchemaDef };
          optional?: infer Optional;
          list?: infer List;
          description?: string;
        }
      ? {
          type: 'schema';
          optional: Optional extends true ? true : false;
          list: List extends true ? true : false;
          def: SchemaDef;
          description?: string;
        }
      : // end 'schema'

      // record start
      T extends {
          type: 'record';
          optional?: infer Optional;
          list?: infer List;
          def?: infer Def;
          description?: string;
        }
      ? {
          type: 'record';
          optional: [Optional] extends [true] ? true : false;
          list: [List] extends [true] ? true : false;
          def: {
            keyType: Def extends { keyType: 'int' | 'float' } ? Def['keyType'] : 'string';
            type: Def extends { type: FieldDefinitionConfig } ? Def['type'] : 'any';
          };
          description?: string;
        }
      : // record end

      T extends {
          type: FieldTypeName;
          optional?: infer Optional;
          list?: infer List;
          def?: infer Def;
          description?: string;
        }
      ? {
          type: T['type'];
          optional: Optional extends true ? true : false;
          list: List extends true ? true : false;
          def: OnlyKnown<Def>;
          description?: string;
        }
      : T extends FieldType<any, infer TypeName, infer Def>
      ? {
          type: TypeName;
          optional: T['isOptional'] extends true ? true : false;
          list: T['isList'] extends true ? true : false;
          def: Def;
          description?: string;
        }
      : T extends AnyStringFieldDefinition
      ? ParseStringDefinition<T>
      : // ======= H start ====
      [IsFieldAsSingleKey<T>] extends [true]
      ? ParseFieldNameAsSingleKey<T>
      : // ======= H end ====
        never
  ) extends infer Parsed
    ? Parsed extends Record<string, any>
      ? {
          // fix 'boolean' is not assignable to type 'true' 😔
          [K in keyof Parsed]: K extends 'optional' ? ([Parsed[K]] extends [never] ? false : Parsed[K]) : Parsed[K];
        }
      : never
    : never;

type NullableKeys<T extends ParsedSchemaDefinition<any>, Kind extends 'nullable' | 'required'> = {
  [K in keyof T]?: T[K]['optional'] extends (Kind extends 'nullable' ? true : false) ? K : never;
}[keyof T] extends infer Temp
  ? Temp extends undefined
    ? never
    : Extract<Temp, string>
  : never;

export type TypeFromSchemaDefinition<Schema extends ParsedSchemaDefinition<any>> = {
  [K in keyof ({
    [K in NullableKeys<Schema, 'required'>]: TypeFromSchema<Schema[K]>;
  } & {
    [K in NullableKeys<Schema, 'nullable'>]?: TypeFromSchema<Schema[K]>;
  })]: ({
    [K in NullableKeys<Schema, 'nullable'>]?: TypeFromSchema<Schema[K]>;
  } & {
    [K in NullableKeys<Schema, 'required'>]: TypeFromSchema<Schema[K]>;
  })[K];
};

// === START to infer union as optional ===
export type HasNullableDefinition<T> =
  | IsOptionalDef<T>
  | IsOptionalFieldInstance<T>
  | IsOptionalString<T>
  | IsOptionalUnion<T>;

type IsOptionalString<T> = T extends string ? (T extends `${string}?` ? true : never) : never;
type IsOptionalDef<T> = T extends object ? (T extends { optional: true } ? true : never) : never;
type IsOptionalFieldInstance<T> = T extends object ? (T extends { isOptional: true } ? true : never) : never;
type IsOptionalUnion<T> = T extends object
  ? T extends FieldType<any, any, readonly (infer Def)[]>
    ? HasNullableDefinition<Def>
    : never
  : never;
// === END to infer union as optional ===

export type IsFieldAsSingleKey<T> = T extends { [K in keyof FieldTypes]?: any }
  ? [Exclude<keyof T, keyof CommonFieldConfig | keyof FieldTypes>] extends [never]
    ? true
    : never
  : never;

export type ParseFieldNameAsSingleKey<T> =
  //
  T extends { [K in keyof FieldTypes]?: any }
    ? [Exclude<keyof T, keyof CommonFieldConfig | keyof FieldTypes>] extends [never]
      ? {
          type: Exclude<keyof T, keyof CommonFieldConfig>;
          def: T[Exclude<keyof T, keyof CommonFieldConfig>];
          optional: T extends { optional: true } ? true : false;
          list: T extends { list: true } ? true : false;
          description?: string;
        }
      : never
    : never;
