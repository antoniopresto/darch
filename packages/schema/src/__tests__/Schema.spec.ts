import { assert, IsExact } from 'conditional-type-checks';

import { createSchema, Schema } from '../Schema';
import { TypeFromSchema } from '../TSchemaParser';

const userSchema = new Schema({
  name: 'string',
  optional: 'string?',
  age: 'int',
  gender: {
    type: 'enum',
    def: ['male', 'female', 'other'],
    optional: true,
  },

  category: ['general', 'closed'],
  '12Enum': ['1', '2'],

  enumArray: {
    type: 'enum',
    list: true,
    def: ['1', '2'],
  },
} as const);

describe('Schema', () => {
  beforeEach(() => {
    Schema.register.clear();
  });

  it('handle definition', () => {
    const sut = userSchema.definition;

    expect(sut).toEqual({
      age: {
        list: false,
        optional: false,
        type: 'int',
      },
      category: {
        def: ['general', 'closed'],
        type: 'enum',
        list: false,
        optional: false,
      },
      '12Enum': {
        def: ['1', '2'],
        type: 'enum',
        list: false,
        optional: false,
      },
      enumArray: {
        def: ['1', '2'],
        type: 'enum',
        list: true,
        optional: false,
      },
      gender: {
        optional: true,
        def: ['male', 'female', 'other'],
        type: 'enum',
        list: false,
      },
      name: {
        list: false,
        optional: false,
        type: 'string',
      },
      optional: {
        list: false,
        optional: true,
        type: 'string',
      },
    });
  });

  it('should validate schema', () => {
    const user = {
      name: 'Antonio',
      age: 1,
      optional: undefined,
      gender: 'male',
      category: 'general',
      '12Enum': '1',
      enumArray: ['1', '2'],
    };

    expect(userSchema.parse(user)).toEqual(user);

    expect(() => userSchema.parse({ ...user, age: undefined })).toThrow(
      `➤ field "age": expected type int, found undefined.`
    );

    expect(() => userSchema.parse({ ...user, enumArray: ['3'] })).toThrow(
      "• enumArray[0] accepted: '1' or '2', found 3."
    );
  });

  it('should validate schema inside schema', () => {
    const subSchema = new Schema({
      mySubField: ['foo'],
    });

    const schema = new Schema({
      name: 'string',
      sub: subSchema,
    });

    expect(
      schema.parse({
        name: 'a',
        sub: { mySubField: 'foo' },
      })
    ).toEqual({ name: 'a', sub: { mySubField: 'foo' } });

    expect(() =>
      schema.parse({
        name: 'a',
        sub: { mySubField: 'INVALID' },
      })
    ).toThrow('➤ field "sub": ➤ field "mySubField": accepted: \'foo\', found INVALID.');

    expect(() => schema.parse({ name: 'a', sub: 1 })).toThrow('➤ field "sub": expected object, found number.');

    expect(() => schema.parse({ name: 'a', sub: {} })).toThrow(
      '➤ field "sub": ➤ field "mySubField": expected type enum, found undefined.'
    );
  });

  it('accepts schema as field type', () => {
    const rolesSchema = new Schema({
      name: 'string',
      permissions: '[string]',
      status: ['open', 'closed'],
    } as const);

    const mySchema = new Schema({
      userId: 'string',
      roles: {
        type: rolesSchema,
        list: true,
      },
    });

    expect(mySchema.definition.roles.type).toBe('schema');
    expect(mySchema.definition.roles.def).toEqual(rolesSchema.definition);

    expect(() => mySchema.parse({ userId: '123' })).toThrow('➤ field "roles": expected type schema, found undefined.');

    expect(() => mySchema.parse({ userId: '123', roles: [] })).not.toThrow();

    expect(() => mySchema.parse({ userId: '123', roles: [1] })).toThrow('• roles[0] expected object, found number');
  });

  test('describe', () => {
    const schema = createSchema({
      name: 'string',
      age: 'int?',
    })
      .describe('my schema desc.')
      .describe({
        name: 'the name field',
        age: 'the age field',
      });

    expect(schema.definition.name.description).toEqual('the name field');
    expect(schema.definition.age.description).toEqual('the age field');
  });

  test('clone', () => {
    const schema1 = createSchema({
      name: 'string',
      age: 'int?',
    });

    const schema2 = schema1.clone();

    expect(schema1.definition).not.toBe(schema2.definition);
    expect(schema1.definition).toEqual(schema2.definition);
    expect(schema1.definition).toEqual(schema2.definition);
  });

  test('removeField', () => {
    const schema1 = createSchema({
      name: 'string',
      age: 'int?',
      email: 'email',
    });

    const noName = schema1.removeField('name');
    const noEmail = schema1.removeField(['email']);

    expect(schema1.definition).toEqual({
      age: {
        list: false,
        optional: true,
        type: 'int',
      },
      name: {
        list: false,
        optional: false,
        type: 'string',
      },
      email: {
        list: false,
        optional: false,
        type: 'email',
      },
    });

    expect(noName.definition).toEqual({
      age: {
        list: false,
        optional: true,
        type: 'int',
      },
      email: {
        list: false,
        optional: false,
        type: 'email',
      },
    });

    type T1 = TypeFromSchema<typeof schema1>;
    type NoName = TypeFromSchema<typeof noName>;
    type NoEmail = TypeFromSchema<typeof noEmail>;

    assert<IsExact<NoName, Omit<T1, 'name'>>>(true);
    assert<IsExact<NoEmail, Omit<T1, 'email'>>>(true);
  });

  test('addFields', () => {
    const schema1 = createSchema({
      name: 'string',
      age: 'int?',
    });

    const withEmail = schema1.addFields({
      email: 'email',
    });

    expect(schema1.definition).toEqual({
      age: {
        list: false,
        optional: true,
        type: 'int',
      },
      name: {
        list: false,
        optional: false,
        type: 'string',
      },
    });

    expect(withEmail.definition).toEqual({
      age: {
        list: false,
        optional: true,
        type: 'int',
      },
      email: {
        list: false,
        optional: false,
        type: 'email',
      },
      name: {
        list: false,
        optional: false,
        type: 'string',
      },
    });

    type T1 = TypeFromSchema<typeof schema1>;
    type WithEmail = TypeFromSchema<typeof withEmail>;

    assert<IsExact<WithEmail, T1 & { email: string }>>(true);
  });

  test('number as string input', () => {
    expect(
      createSchema({
        age: 'int',
        weight: 'float',
      }).parse({ age: '32', weight: '83' })
    ).toEqual({
      age: 32,
      weight: 83,
    });
  });

  test('identify', () => {
    const schema1 = createSchema({
      name: 'string',
      age: 'int?',
    });

    expect(schema1.id).toBe(null);

    // @ts-ignore
    expect(() => schema1.identify()).toThrow();
    expect(() => schema1.identify('')).toThrow();

    expect(schema1.identify('abc')).toHaveProperty('id', 'abc');
    schema1.identify('abc')
    schema1.identify('abc')
    expect(() => schema1.identify('abcx')).toThrow('Trying to replace existing id "abc"');

    expect(Schema.register.get('abc')).toBe(schema1);
    expect(() => Schema.register.get('yyy')).toThrow('There is no item with key "yyy"');
  });

  test('.graphqlType()', () => {
    const type = createSchema({
      name: 'string',
      age: 'int?',
    }).identify('User');

    expect(type.graphqlType().getFields()).toMatchSnapshot();
  });

  test('.graphqlInputType()', () => {
    const type = createSchema({
      name: 'string',
      age: 'int?',
    }).identify('User');

    expect(type.graphqlInputType().getFields()).toMatchSnapshot();
  });
});
