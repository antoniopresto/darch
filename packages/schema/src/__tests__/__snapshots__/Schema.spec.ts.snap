// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema .graphqlInputType() 1`] = `
Object {
  "age": Object {
    "astNode": Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": undefined,
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "age",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "Int",
        },
      },
    },
    "description": undefined,
    "name": "age",
    "type": "Int",
  },
  "name": Object {
    "astNode": Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": undefined,
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "name",
      },
      "type": Object {
        "kind": "NonNullType",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "value": "String",
          },
        },
      },
    },
    "description": undefined,
    "name": "name",
    "type": "String!",
  },
}
`;

exports[`Schema .graphqlType() 1`] = `
Object {
  "age": Object {
    "args": Array [],
    "astNode": Object {
      "arguments": undefined,
      "description": undefined,
      "directives": undefined,
      "kind": "FieldDefinition",
      "name": Object {
        "kind": "Name",
        "value": "age",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "Int",
        },
      },
    },
    "isDeprecated": false,
    "name": "age",
    "type": "Int",
  },
  "name": Object {
    "args": Array [],
    "astNode": Object {
      "arguments": undefined,
      "description": undefined,
      "directives": undefined,
      "kind": "FieldDefinition",
      "name": Object {
        "kind": "Name",
        "value": "name",
      },
      "type": Object {
        "kind": "NonNullType",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "value": "String",
          },
        },
      },
    },
    "isDeprecated": false,
    "name": "name",
    "type": "String!",
  },
}
`;
